pyomo/core/base/matrix_constraint.py

```diff
@@ -0,0 +1,360 @@
+        # types that build the body expression on the
+        # fly do not have to here
+        body = self(exception=False)
+        if body is None:
+            return None
+        ub = self.upper
+        if ub is None:
+            ub = float('inf')
+        else:
+            ub = value(ub)
+        return ub - body
+
+    def slack(self):
+        """min(lslack, uslack). Returns :const:`None` if a
+        value for the body can not be computed."""
+        # this method is written so that constraint
+        # types that build the body expression on the
+        # fly do not have to here
+        body = self(exception=False)
+        if body is None:
+            return None
+        return min(self.lslack(), self.uslack())
+
+    #
+    # Override some default implementations on ComponentData
+    #
+
+    def index(self):
+        return self._index
+
+    #
+    # Abstract Interface (_ConstraintData)
+    #
+
+    @property
+    def body(self):
+        """Access the body of a constraint expression."""
+        comp = self.parent_component()
+        index = self._index
+        data = comp._A_data
+        indices = comp._A_indices
+        indptr = comp._A_indptr
+        x = comp._x
+        ptrs = xrange(indptr[index],
+                      indptr[index+1])
+        return LinearExpression(
+            linear_vars=[x[indices[p]] for p in ptrs],
+            linear_coefs=[data[p] for p in ptrs],
+            constant=0
+        )
+
+    @property
+    def lower(self):
+        """Access the lower bound of a constraint
+        expression."""
+        comp = self.parent_component()
+        index = self._index
+        return comp._lower[index]
+
+    @property
+    def upper(self):
+        """Access the upper bound of a constraint
+        expression."""
+        comp = self.parent_component()
+        index = self._index
+        return comp._upper[index]
+
+    @property
+    def equality(self):
+        """A boolean indicating whether this is an equality
+        constraint."""
+        comp = self.parent_component()
+        index = self._index
+        if (comp._lower[index] is None) or \
+           (comp._upper[index] is None):
+            return False
+        return comp._lower[index] == comp._upper[index]
+
+    @property
+    def strict_lower(self):
+        """A boolean indicating whether this constraint has
+        a strict lower bound."""
+        return False
+
+    @property
+    def strict_upper(self):
+        """A boolean indicating whether this constraint has
+        a strict upper bound."""
+        return False
+
+    def set_value(self, expr):
+        """Set the expression on this constraint."""
+        raise NotImplementedError(
+            "MatrixConstraint row elements can not be updated"
+        )
+
+@ModelComponentFactory.register(
+                   "A set of constraint expressions in Ax=b form.")
+class MatrixConstraint(collections_Mapping,
+                       IndexedConstraint):
```

Comment:
Can you comment on why you inherit from both collections_Mapping and IndexedConstraint?  In particular, you are having the collections API overwrite the IndexedConstraint API.  What "weirdness" are you attempting to circumvent?  If there are efficiency or implementational benefits of doing this, we should consider having IndexedComponent inherit from collections_Mapping...

Reply:
This may just be a remnant of the implementation in kernel. I'll have to double check.